### ISSUE TITLE
[REVIEW]: GXBeam: A Pure Julia Implementation of Geometrically Exact Beam Theory

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3997]
**Submitting author:** @taylormcd (<a href="http://orcid.org/0000-0001-7993-5703">Taylor McDonnell</a>)
**Repository:** <a href="https://github.com/byuflowlab/GXBeam.jl" target ="_blank">https://github.com/byuflowlab/GXBeam.jl</a>
**Version:** v0.3.1
**Editor:** @prashjha
**Reviewer:** @Kevin-Mattheus-Moerman, @PetrKryslUCSD
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/13cb0c41e9834510c6acf732bdfa8c05/status.svg)](https://joss.theoj.org/papers/13cb0c41e9834510c6acf732bdfa8c05)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/13cb0c41e9834510c6acf732bdfa8c05"><img src="https://joss.theoj.org/papers/13cb0c41e9834510c6acf732bdfa8c05/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/13cb0c41e9834510c6acf732bdfa8c05/status.svg)](https://joss.theoj.org/papers/13cb0c41e9834510c6acf732bdfa8c05)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/byuflowlab/GXBeam.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@Kevin-Mattheus-Moerman & @PetrKryslUCSD, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @prashjha know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @Kevin-Mattheus-Moerman
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/byuflowlab/GXBeam.jl">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/byuflowlab/GXBeam.jl">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@taylormcd) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @PetrKryslUCSD
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/byuflowlab/GXBeam.jl">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/byuflowlab/GXBeam.jl">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@taylormcd) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @Kevin-Mattheus-Moerman, @PetrKryslUCSD it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 822

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.09 s (386.2 files/s, 191412.8 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Julia                           13           1729           1069           7310
Markdown                         7            751              0           3076
SVG                              6              6              6           2656
TeX                              1             10              0            119
YAML                             4              5              0             71
TOML                             3              2              0             41
-------------------------------------------------------------------------------
SUM:                            34           2503           1075          13273
-------------------------------------------------------------------------------


Statistical information for the repository 'bef385b8dd4c87a3022f9756' was
gathered on 2021/12/17.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.5281/zenodo.4404703 is OK
- 10.5281/zenodo.5633646 is OK
- 10.5334/jors.151 is OK
- 10.2514/5.9781600866821.0103.0142 is OK
- 10.1002/sapm197352287 is OK
- 10.1016/j.compstruct.2012.04.007 is OK
- 10.1063/1.4985091 is OK
- 10.2514/5.9781600866821.0059.0102 is OK

MISSING DOIs

- 10.1137/141000671 may be a valid DOI for title: Julia: A fresh approach to numerical computing

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03997/joss.03997/10.21105.joss.03997.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03997/joss.03997/10.21105.joss.03997.pdf) :page_facing_up: :point_left:

**Comment by prashjha** [comment]
Thank you @Kevin-Mattheus-Moerman and @PetrKryslUCSD for agreeing to be reviewers. Please read the first couple of comments in this thread and also reviewer guidelines [here](https://joss.readthedocs.io/en/latest/reviewer_guidelines.html). Also, you can browse the closed "REVIEW" issues on the "joss-reviews" repository to get some ideas on how to complete the reviews. Good luck!

**Comment by whedon** [comment]
:wave: @PetrKryslUCSD, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @Kevin-Mattheus-Moerman, please update us on how your review is going (this is an automated reminder).

**Comment by PetrKryslUCSD** [comment]
Waiting on responses to my suggestions/questions.

**Comment by taylormcd** [comment]
Sorry for the delay.  Had the holidays, then a conference, and am now dealing with being sick.  I should be able to address suggestions/questions starting next week.

**Comment by PetrKryslUCSD** [comment]
No worries.

**Comment by PetrKryslUCSD** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03997/joss.03997/10.21105.joss.03997.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03997/joss.03997/10.21105.joss.03997.pdf) :page_facing_up: :point_left:

**Comment by taylormcd** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03997/joss.03997/10.21105.joss.03997.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03997/joss.03997/10.21105.joss.03997.pdf) :page_facing_up: :point_left:

**Comment by PetrKryslUCSD** [comment]
@whedon help

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by PetrKryslUCSD** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by PetrKryslUCSD** [comment]
@prashjha I am sorry, I can't figure out how to submit the review. Is there a whedon command?

**Comment by prashjha** [comment]
Hi @PetrKryslUCSD, to review, you need to look at the checklists provided, open issues in the software repository to fix issues that you have in mind, and add comments here to the corresponding authors if any. Once your issues have been addressed and you are satisfied with the response and overall quality of the submission, you will simply let me know your decision here.

**Comment by PetrKryslUCSD** [comment]
OK. In that case I would like to report the review is finished. Recommendation: publish.

**Comment by prashjha** [comment]
@PetrKryslUCSD, got it. Thank you for taking the time to review this.

**Comment by Kevin-Mattheus-Moerman** [comment]
@prashjha I'll get to my review in the coming week. Apologies for the delay so far.

**Comment by Kevin-Mattheus-Moerman** [comment]
@taylormcd I've started the review process. Here are some initial comments. I'll resume work early next week. 

- [ ] On the [documentation site]() under `Limitations` it says `..More details about the convergence of this package may be found in the examples...`. The example link there is broken (it goes here which gives a 404: https://flow.byu.edu/GXBeam.jl/dev/@ref%20Nonlinear%20Analysis%20of%20a%20Cantilever%20Subjected%20to%20a%20Constant%20Moment)
- [ ] The examples in the documentation appear encoded in a mardown file (i.e. https://github.com/byuflowlab/GXBeam.jl/blob/master/docs/src/examples.md). It would be great however if these could be executed e.g. as jupyter notebooks. Could the authors consider facilitating this? It should largely be a copy-paste excercise to make these executable Jupyter notebooks, whereas currently, as a new user, I find myself having to copy paste a lot to reproduce the examples. 
- [ ] I got an error running the tests, can you assist/investigate? https://github.com/byuflowlab/GXBeam.jl/issues/68

**Comment by taylormcd** [comment]
@Kevin-Mattheus-Moerman 
 - The broken link is fixed now.
 - I used ![Documenter.jl](https://github.com/JuliaDocs/Documenter.jl) for the documentation and examples, which is pretty standard practice in Julia.  I would rather not have two versions of the same examples due to the headaches that causes for maintenance.  That being said, using ![Literate.jl](https://github.com/fredrikekre/Literate.jl) might be a good long term solution.
 - The error is now fixed.

**Comment by taylormcd** [comment]
@Kevin-Mattheus-Moerman 
Jupyter notebook versions of the examples are now available.

**Comment by Kevin-Mattheus-Moerman** [comment]
@taylormcd thanks for adding those. I'll have a look later the coming week. Where can I find the new Jupyter notebook versions?

**Comment by taylormcd** [comment]
They're linked from the relevant pages in the documentation. Note that I haven't yet created a new release with the changes, so the notebooks are only linked from the development version of the documentation..

**Comment by prashjha** [comment]
Hi @Kevin-Mattheus-Moerman, just checking if you are close to finishing this review. No worries if you need further time.

**Comment by Kevin-Mattheus-Moerman** [comment]
@taylormcd 

- [ ] I get some errors during testing, see; https://github.com/byuflowlab/GXBeam.jl/issues/77

**Comment by taylormcd** [comment]
I added a short contributing section to README.md.  I'll probably create a more detailed one later, but this should suffice for now.

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot check references

**Comment by Kevin-Mattheus-Moerman** [comment]
@editorialbot generate pdf

**Comment by Kevin-Mattheus-Moerman** [comment]
@taylormcd @prashjha all issues have now been addressed and I recommend acceptance in JOSS. I wanted to check references, see above, however it seems our bot is not responding to references checking requests this morning. But I am sure you'll check this during the potential acceptance process. Thanks for the opportunity to review this interesting package. All the best.

**Comment by prashjha** [comment]
@editorialbot check references

**Comment by prashjha** [comment]
Hi, @Kevin-Mattheus-Moerman, thanks for completing the review.

**Comment by prashjha** [comment]
Hello @taylormcd, I am reading the draft one last time. If I have any edits or suggestions, I will let you know soon. Meanwhile, could you check the following reference flagged by editorialbot:

Ibrahimbegović, A., & Mikdad, M. A. (1998). Finite rotations in dynamics of beams and implicit time-stepping schemes. International Journal for Numerical Methods in Engineer- ing, 41(5), 781–814. https://doi.org/10.1002/(SICI)1097-0207(19980315)41:5%3C781:: AID-NME308%3E3.0.CO;2-9

**Comment by prashjha** [comment]
Hi, @taylormcd, see if these suggestions for the draft makes sense:

- Line 16: missing space between "researchers" and "(Betsch".
- Line 33: "that it is written ..."?
- Line 53: "... accelerations are also provided" is better as you use word "support" in the beginning of this sentence.
- Line 65: "... sectional analysis"?

After you have updated the draft and also fixed the reference I mentiond, can you also do (if not done already) a 'tagged' release of your code, archive the release using zenedo or other methods? Make sure that the title of zenedo archive matches with the title of this JOSS submission.

Once you are done, I will hand your paper to EiC for the final decision.

**Comment by taylormcd** [comment]
I adopted the first three changes, but not the last one because multiple section analyses may be necessary for a single GXBeam run.  I also tagged a new release and archived it with Zenodo: https://doi.org/10.5281/zenodo.6558458.  Let me know if there's anything else that needs to be done.

**Comment by prashjha** [comment]
@editorialbot generate pdf

**Comment by prashjha** [comment]
@editorialbot recommend-accept

**Comment by prashjha** [comment]
@taylormcd, compiler is not happy with bib entries (Bezanson2017 and Hodges2006) in paper.bib file. Compiler expects issn value to have exactly 9 elements (e.g., xxxx-xxxx). Could you check these two bib entries?

**Comment by arfon** [comment]
Based on past experience, these may in fact be ISBNs rather than ISSN entries.

**Comment by taylormcd** [comment]
Bezanson2017 has two ISSN: 0036-1445 and 1095-7200, so I didn't change anything there.  For Hodges, it was a ISBN rather than an ISSN, and the issue is now corrected.

**Comment by prashjha** [comment]
@editorialbot generate pdf

**Comment by prashjha** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@editorialbot accept
